nameserver/server.go

```diff
@@ -32,7 +20,93 @@ func makeDNSFailResponse(r *dns.Msg) *dns.Msg {
 	return m
 }
 
-func queryHandler(lookups []Lookup) dns.HandlerFunc {
+type DNSServer struct {
+	config  *dns.ClientConfig
+	zone    Zone
+	iface   *net.Interface
+	udpPort int
+	udpSrv  *dns.Server
+	tcpPort int
+	tcpSrv  *dns.Server
+}
+
+// Creates a new DNS server with a given config
+func NewDNSServerWithConfig(config *dns.ClientConfig, zone Zone, iface *net.Interface, udpPort int, tcpPort int) (*DNSServer, error) {
+	return &DNSServer{
+		config:  config,
+		zone:    zone,
+		iface:   iface,
+		udpPort: udpPort,
+		tcpPort: tcpPort,
+	}, nil
+}
+
+// Creates a new DNS server
+func NewDNSServer(zone Zone, iface *net.Interface, udpPort int, tcpPort int) (*DNSServer, error) {
+	config, err := dns.ClientConfigFromFile("/etc/resolv.conf")
+	if err != nil {
+		return nil, err
+	}
+
+	return NewDNSServerWithConfig(config, zone, iface, udpPort, tcpPort)
+}
+
+// Start the DNS server
+func (s *DNSServer) Start() error {
+	mdnsClient, err := NewMDNSClient()
+	checkFatal(err)
+
+	ifaceName := "default interface"
+	if s.iface != nil {
+		ifaceName = s.iface.Name
+	}
+	Info.Printf("Using mDNS on %s", ifaceName)
+	err = mdnsClient.Start(s.iface)
+	checkFatal(err)
+
+	LocalServeMux := dns.NewServeMux()
+	LocalServeMux.HandleFunc(LOCAL_DOMAIN, s.queryHandler([]Lookup{s.zone, mdnsClient}))
+	LocalServeMux.HandleFunc(RDNS_DOMAIN, s.rdnsHandler([]Lookup{s.zone, mdnsClient}))
+	LocalServeMux.HandleFunc(".", s.notUsHandler())
+
+	mdnsServer, err := NewMDNSServer(s.zone)
+	checkFatal(err)
+
+	err = mdnsServer.Start(s.iface)
+	checkFatal(err)
+
+	udpAddress := fmt.Sprintf(":%d", s.udpPort)
+	s.udpSrv = &dns.Server{Addr: udpAddress, Net: "udp", Handler: LocalServeMux}
+	tcpAddress := fmt.Sprintf(":%d", s.tcpPort)
+	s.tcpSrv = &dns.Server{Addr: tcpAddress, Net: "tcp", Handler: LocalServeMux}
+
+	go func() {
+		Info.Printf("Listening for DNS on %s (UDP)", udpAddress)
+		err = s.udpSrv.ListenAndServe()
+		checkFatal(err)
+	}()
+
+	go func() {
+		Info.Printf("Listening for DNS on %s (TCP)", tcpAddress)
+		err = s.tcpSrv.ListenAndServe()
+		checkFatal(err)
+	}()
+
+	return nil
+}
+
+// Perform a graceful shutdown
+func (s *DNSServer) Stop() error {
+	if err := s.tcpSrv.Shutdown(); err != nil {
+		return err
+	}
+	if err := s.udpSrv.Shutdown(); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (s *DNSServer) queryHandler(lookups []Lookup) dns.HandlerFunc {
```

Comment:
Oh it is by analogy with the other *Handler functions. I think there is a better arrangement in there somewhere, but this is OK for now.